### PR TITLE
Support MSI resource id authentication

### PR DESCRIPTION
**Pull request by serbrech**
allows to use msi resource id instead of msi client id for authentication

- updates autorest/adal
- deprecate funcs that did not make the difference between clientid and resourceid

### CODE REVIEW

File: internal/aad/jwt.go [hunk 1]
```diff
@@ -201,10 +228,13 @@ func (c *TokenProviderConfiguration) NewServicePrincipalToken() (*adal.ServicePr
...
-	if c.ClientID == "" {
-		return adal.NewServicePrincipalTokenFromMSI(msiEndpoint, c.ResourceURI)
+	if c.ClientID != "" {
+		return adal.NewServicePrincipalTokenFromMSIWithUserAssignedID(msiEndpoint, c.ResourceURI, c.ClientID)
+	}
+	if c.ResourceID != "" {
+		return adal.NewServicePrincipalTokenFromMSIWithIdentityResourceID(msiEndpoint, c.ResourceURI, c.ResourceID)
 	}
-	return adal.NewServicePrincipalTokenFromMSIWithUserAssignedID(msiEndpoint, c.ResourceURI, c.ClientID)
+	return adal.NewServicePrincipalTokenFromMSI(msiEndpoint, c.ResourceURI)
```
Comment (keikumata) [inline internal/aad/jwt.go:237]: doesn't this default to using the system assigned managed identity?

File: go.mod [hunk 2]
```diff
@@ -6,9 +6,9 @@ require (
 	github.com/Azure/azure-amqp-common-go/v3 v3.0.0
 	github.com/Azure/azure-sdk-for-go v37.1.0+incompatible
 	github.com/Azure/azure-service-bus-go v0.10.1
-	github.com/Azure/go-autorest/autorest v0.9.3
-	github.com/Azure/go-autorest/autorest/adal v0.8.1
+	github.com/Azure/go-autorest/autorest v0.11.6
```
Comment (serbrech) [inline go.mod:9]: bumping to latest adal that added support for msi_rid IMDS authentication

File: internal/aad/jwt.go [hunk 3]
```diff
@@ -88,11 +89,28 @@ func JWTProviderWithClientCertificate(clientID, certificatePath, certificatePass
...
 }
 
-// JWTProviderWithManagedIdentity configures the TokenProvider using managed identity
-// - attempt to authenticate using managed identity. If clientID is supplied then it will use that
-//   as the clientID of the user assigned managed identity. Otherwise it will default to the system assigned managed
+// JWTProviderWithManagedIdentityResourceID configures the TokenProvider using managed identity
+// - attempt to authenticate using managed identity resourceID. Otherwise it will default to the system assigned managed
 //   identity
-func JWTProviderWithManagedIdentity(clientID, environment string) JWTProviderOption {
+func JWTProviderWithManagedIdentityResourceID(resourceID, environment string) JWTProviderOption {
```
Comment (serbrech) [inline internal/aad/jwt.go:95]: adding ResourceID JWTProviderOption

**Review by keikumata — APPROVED** [review]
LGTM

File: internal/servicebus/namespace.go [hunk 4]
```diff
@@ -3,32 +3,66 @@ package servicebus
...
 
+	"github.com/Azure/azure-amqp-common-go/v3/auth"
 	servicebussdk "github.com/Azure/azure-service-bus-go"
 
 	"github.com/keikumata/azure-pub-sub/internal/aad"
 )
 
 const (
+	// TODO:take this value from config or environment
 	serviceBusResourceURI = "https://servicebus.azure.net/"
```
Comment (serbrech) [inline internal/servicebus/namespace.go:14]: this works for AzurePublicCloud but will need to be different for other cloud (Mooncake, for example)

**Comment by keikumata** [comment]
@serbrech could you add the `NamespaceWithManagedIdentityResourceID` and `NamespaceWithManagedIdentityClientID` to the README?

File: internal/servicebus/namespace.go [hunk 5]
```diff
@@ -3,32 +3,66 @@ package servicebus
...
 )
 
+var errorOption func(h *servicebussdk.Namespace) error
```
Comment (serbrech) [inline internal/servicebus/namespace.go:17]: just an alias to clean up the use case

File: internal/aad/jwt.go [hunk 6]
```diff
@@ -107,6 +125,15 @@ func JWTProviderWithManagedIdentity(clientID, environment string) JWTProviderOpt
 	}
 }
 
+// JWTProviderWithManagedIdentity configures the TokenProvider using managed identity
+// - attempt to authenticate using managed identity. If clientID is supplied then it will use that
+//   as the clientID of the user assigned managed identity. Otherwise it will default to the system assigned managed
+//   identity
+// Deprecated use JWTProviderWithManagedIdentityClientID
+func JWTProviderWithManagedIdentity(clientID, environment string) JWTProviderOption {
```
Comment (serbrech) [inline internal/aad/jwt.go:133]: Deprecated the unspecific signature. use `JWTProviderWithManagedIdentityClientID` or `JWTProviderWithManagedIdentityResourceID` instead

File: internal/aad/jwt.go [hunk 7]
```diff
@@ -200,11 +227,14 @@ func (c *TokenProviderConfiguration) NewServicePrincipalToken() (*adal.ServicePr
 	return spToken, nil
 }
 
-func (c *TokenProviderConfiguration) getTokenProvider(msiEndpoint string) (*adal.ServicePrincipalToken, error){
-	if c.ClientID == "" {
-		return adal.NewServicePrincipalTokenFromMSI(msiEndpoint, c.ResourceURI)
+func (c *TokenProviderConfiguration) getTokenProvider(msiEndpoint string) (*adal.ServicePrincipalToken, error) {
```
Comment (serbrech) [inline internal/aad/jwt.go:230]: if we have clientID use that.
then if we have resourceID use that
then if we have nothing, assume we are using SystemAssigned

File: internal/servicebus/namespace.go [hunk 8]
```diff
@@ -3,32 +3,66 @@ package servicebus
...
+
+func newErrorOption(err error) servicebussdk.NamespaceOption {
+	return func(ns *servicebussdk.Namespace) error { return err }
+}
+
 // NamespaceWithManagedIdentity is a custom NamespaceOption to instantiate a Service Bus namespace client with
-// managed identity
-func NamespaceWithManagedIdentity(serviceBusNamespaceName, managedIdentityClientID string) servicebussdk.NamespaceOption {
+// managed identity resource id
+func NamespaceWithManagedIdentityResourceID(serviceBusNamespaceName, managedIdentityResourceID string) servicebussdk.NamespaceOption {
```
Comment (serbrech) [inline internal/servicebus/namespace.go:25]: Same wrapper around the JWTtoken options

File: listener.go [hunk 9]
```diff
@@ -46,13 +46,33 @@ func ListenerWithConnectionString(connStr string) ListenerManagementOption {
 	}
 }
 
-// ListenerWithManagedIdentity configures a listener with the attached managed identity and the Service bus resource name
+// Deprecated use ListenerWithManagedIdentityClientID instead
```
Comment (serbrech) [inline listener.go:49]: deprecate unspecific listener creator, add ClientID and ResourceID funcs instead

File: README.md [hunk 10]
```diff
@@ -87,23 +91,43 @@ publisher, _ := pubsub.NewPublisher(
...
 
 #### Using system assigned managed identity
+Using Identity ClientID
 ```
 // keep managed identity client ID empty to default using system assigned managed identity
 topicName := "topic"
 publisher, _ := pubsub.NewPublisher(
     topicName,
-    pubsub.PublisherWithManagedIdentity(serviceBusNamespaceName, ""),
+    pubsub.PublisherWithManagedIdentityClientID(serviceBusNamespaceName, identityClientID),
```
Comment (xhl873) [inline README.md:118]: should this be "" for system managed identity?

Comment (serbrech) [reply README.md:118]: yes. for clientID you specify it. if you pass empty, then it uses SystemIdentity.
the comment just says that if you keep it empty, that's what happens.

**Pull request merged**